torch/csrc/jit/python/python_sugared_value.cpp

```diff
@@ -217,6 +217,54 @@ std::shared_ptr<SugaredValue> PythonModuleValue::attr(
   return toSugaredValue(member, m, loc, /*is_constant=*/true);
 }
 
+ModuleValue::ModuleValue(
+    Value* self,
+    std::shared_ptr<ConcreteModuleType> concreteType)
+    : self_(self), concreteType_(std::move(concreteType)) {
+  auto containedTypeHint = concreteType_->getContainedTypeHint();
+
+  if (containedTypeHint) {
+    // For now, only dict type containedTypeHints are supported. Generate and
+    // emit the dictionary representing the ModuleDict into the graph once so
+    // that it can be reused across lookups.
+    DictTypePtr dict_type = containedTypeHint->expect<DictType>();
+    auto* graph = self_->owningGraph();
+    std::vector<Value*> keys, values;
+
+    // Gather submodule names.
+    std::vector<std::string> submodule_names;
+    const auto& self_type = concreteType_->getJitType()->expect<ClassType>();
+    for (size_t i = 0; i < self_type->numAttributes(); ++i) {
+      const auto& attr_type = self_type->getAttribute(i);
+      if (attr_type->is_module()) {
+        if (!attr_type->isSubtypeOf(dict_type->getValueType())) {
+          auto loc = self->node()->sourceRange();
+          throw ErrorReport(loc)
+              << "Attribute " << self_type->getAttributeName(i)
+              << " is not of annotated type "
+              << dict_type->getValueType()->annotation_str();
+        }
+        submodule_names.push_back(self_type->getAttributeName(i));
+      }
+    }
+
+    // Gather Values for the keys (submodule names) and values (submodules) of
+    // this ModuleDict.
+    for (const auto& name : submodule_names) {
+      auto name_v = insertConstant(*graph, name);
+      Value* module_v = graph->insertGetAttr(self_, name);
+      module_v->setType(dict_type->getValueType());
```

Comment:
I don't think we should reset the type, I don't know if that is needed since we're only using it to construct the dictionary. I also dont know if it has any unintended consequences.

Reply:
I was able to remove this now that we're using a special op, but I did need to do something similar when compiling after deserialization. At that time, there are no `ConcreteTypeBuilder` and `ConcreteType` instances, and all modules are compiled as class types. Consequently, the output type of the indexing operation is taken from the `prim::ModuleDictIndex` operator I introduced (which is `Any`, not specific enough), which I reset to the correct interface type.